pkg/controllers/static/helpers.go

```diff
@@ -0,0 +1,158 @@
+) []*scheduling.NodeClaim {
+	nodeClaims := make([]*scheduling.NodeClaim, 0, count)
+	for range count {
+		nct := GetStaticNodeClaimTemplate(np, instanceTypes)
+		nodeClaims = append(nodeClaims, &scheduling.NodeClaim{
+			NodeClaimTemplate: *nct,
+			IsStaticNode:      true,
+		})
+	}
+	return nodeClaims
+}
+
+func GetStaticNodeClaimTemplate(np *v1.NodePool, instanceTypes []*cloudprovider.InstanceType) *scheduling.NodeClaimTemplate {
+	nct := scheduling.NewNodeClaimTemplate(np)
+	nct.InstanceTypeOptions, _, _ = scheduling.FilterInstanceTypesByRequirements(
+		instanceTypes,
+		nct.Requirements,
+		corev1.ResourceList{},
+		corev1.ResourceList{},
+		corev1.ResourceList{},
+		true,
+	)
+	return nct
+}
+
+func ComputeNodeClaimsToProvision(c *state.Cluster, np *v1.NodePool, nodes int64) int64 {
+	limit, ok := np.Spec.Limits[resources.Node]
+	nodeLimit := lo.Ternary(ok, limit.Value(), int64(math.MaxInt64))
+	return c.NodePoolState.ReserveNodeCount(np.Name, nodeLimit, lo.FromPtr(np.Spec.Replicas)-nodes)
+}
+
+func TotalNodesForNodePool(c *state.Cluster, np *v1.NodePool) int64 {
+	running, deleting := c.NodePoolState.GetNodeCount(np.Name)
+	return int64(running + deleting)
+}
+
+// Returns nodes suitable for deprovisioning, prioritizing:
+// 1. Empty nodes (nodes with no pods or only DaemonSet pods)
+// 2. If more nodes needed, nodes with lowest disruption cost
+func GetDeprovisioningCandidates(ctx context.Context, kubeClient client.Client, np *v1.NodePool, nodes []*state.StateNode, count int, clk clock.Clock) []*state.StateNode {
+	// First get empty nodes
+	emptyNodes := lo.Filter(nodes, func(node *state.StateNode, _ int) bool {
+		pods, err := node.Pods(ctx, kubeClient)
+		if err != nil {
+			log.FromContext(ctx).WithValues("node", node.Name()).Error(err, "unable to list pods, treating as non-empty")
+			return false
+		}
+		return len(pods) == 0 || lo.EveryBy(pods, pod.IsOwnedByDaemonSet)
+	})
+
+	candidates := lo.Slice(emptyNodes, 0, count)
+	remaining := count - len(candidates)
+
+	if remaining > 0 {
+		emptyNodesSet := lo.SliceToMap(emptyNodes, func(n *state.StateNode) (*state.StateNode, struct{}) {
+			return n, struct{}{}
+		})
+		nonEmptyNodesWithCost := lo.FilterMap(nodes, func(node *state.StateNode, _ int) (NodeDisruptionCost, bool) {
+			if _, ok := emptyNodesSet[node]; ok {
+				return NodeDisruptionCost{}, false
+			}
+
+			pods, err := node.Pods(ctx, kubeClient)
+			if err != nil {
+				log.FromContext(ctx).WithValues("node", node.Name()).Error(err, "unable to list pods, skipping node")
+				return NodeDisruptionCost{}, false
+			}
+
+			return NodeDisruptionCost{
+				node: node,
+				cost: disruptionutils.ReschedulingCost(ctx, pods) *
+					disruptionutils.LifetimeRemaining(clk, np, node.NodeClaim),
+			}, true
+		})
+
+		// Optimization : Dont need to sort all items, instead use a heap to get remaining
+		sort.Slice(nonEmptyNodesWithCost, func(i, j int) bool {
+			return nonEmptyNodesWithCost[i].cost < nonEmptyNodesWithCost[j].cost
```

Comment:
We should also consider blocked PDBs and pods with the `karpenter.sh/do-not-disrupt` annotation here. These aren't considered in the disruption cost calculation since these pods aren't eligible for consolidation, but they are eligible here. I think you end up with this priority set: `default > blocking PDBs > do-not-disrupt`. Within each set, you still sort by disruption cost.

Reply:
I don't have a strong opinion on PDBs as a follow-up, but definitely think we should deprioritize nodes with pods with the `do-not-disrupt` annotation. The reason being, Karpenter won't evict pods with that annotation. This could leave nodes stuck draining indefinitely.

Reply:
Added